![My Image](https://raw.githubusercontent.com/ralf-42/Image/main/genai-banner-2.jpg)

<p><font size="5" color='grey'> <b>
Multimodales RAG
</b></font> </br></p>


---

In [1]:
#@title 🔧 Umgebung einrichten{ display-mode: "form" }
!uv pip install --system -q git+https://github.com/ralf-42/GenAI.git#subdirectory=04_modul
from genai_lib.utilities import check_environment, get_ipinfo, setup_api_keys, mprint, install_packages
setup_api_keys(['OPENAI_API_KEY', 'HF_TOKEN'], create_globals=False)
print()
check_environment()
print()
get_ipinfo()

✓ OPENAI_API_KEY erfolgreich gesetzt
✓ HF_TOKEN erfolgreich gesetzt

Python Version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]

Installierte LangChain-Bibliotheken:
langchain                                1.0.3
langchain-classic                        1.0.0
langchain-community                      0.4.1
langchain-core                           1.0.2
langchain-experimental                   0.0.42
langchain-ollama                         1.0.0
langchain-openai                         1.0.1
langchain-text-splitters                 1.0.0

IP-Adresse: 35.238.84.201
Hostname: 201.84.238.35.bc.googleusercontent.com
Stadt: Council Bluffs
Region: Iowa
Land: US
Koordinaten: 41.2619,-95.8608
Provider: AS396982 Google LLC
Postleitzahl: 51502
Zeitzone: America/Chicago


In [2]:
#@title 🛠️ Installationen { display-mode: "form" }
install_packages([
    ('markitdown[all]', 'markitdown'),
    'langchain_chroma',
])

🔄 Installiere markitdown[all]...
✅ markitdown[all] erfolgreich installiert und importiert
🔄 Installiere langchain_chroma...
✅ langchain_chroma erfolgreich installiert und importiert


In [3]:
#@title 📂 Dokumente, Bilder, Modul { display-mode: "form" }
!rm -rf files
!mkdir files

# --- Texte
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/biografien_1.txt -o files/biografien_1.txt
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/biografien_2.md -o files/biografien_2.md
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/biografien_3.pdf -o files/biografien_3.pdf
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/biografien_4.docx -o files/biografien_4.docx
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/roboter.txt -o files/roboter.txt

# --- Bilder
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/retro_robot.jpg -o files/retro_robot.jpg
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/hedra_cyborg.png -o files/hedra_cyborg.png
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/apfel.png -o files/apfel.png
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/02%20data/zwei_roboter.png -o zwei_roboter.png

# --- Python-Modul
!curl -L https://raw.githubusercontent.com/ralf-42/GenAI/main/01%20ipynb/multimodal_rag_modul.py -o multimodal_rag_modul.py

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5324  100  5324    0     0  25602      0 --:--:-- --:--:-- --:--:-- 25719
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7868  100  7868    0     0  56752      0 --:--:-- --:--:-- --:--:-- 57014
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 68184  100 68184    0     0   304k      0 --:--:-- --:--:-- --:--:--  305k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 18743  100 18743    0     0   105k      0 --:--:-- --:--:-- --:--:--  105k
  % Total    % Received % Xferd  Average Speed   Tim

# 1 | RAG-Prozess
---

In dem entwickelten **multimodalen Retrieval-Augmented-Generation (RAG)**-System werden sowohl Text- als auch Bilddaten verarbeitet.
Für die **Bildverarbeitung** kommt ein zweistufiger Ansatz zum Einsatz:

1. **Image-Embedding:**
   Das Bild wird in einen hochdimensionalen Vektorraum eingebettet, um visuelle Merkmale wie Formen, Farben und Strukturen numerisch zu repräsentieren.

2. **Textbeschreibung und Text-Embedding:**
   Zusätzlich wird mit *gpt-4o-mini* eine sprachliche Beschreibung des Bildes erzeugt. Diese Beschreibung wird anschließend in ein Text-Embedding überführt, um semantische Informationen textbasiert nutzbar zu machen.

---

**Vorteile des Ansatzes:**

* **Erweiterte semantische Repräsentation:**
  Durch die Kombination von visuellen und sprachlichen Embeddings werden sowohl konkrete als auch konzeptuelle Eigenschaften eines Bildes abgebildet.

* **Verbesserte Retrieval-Qualität:**
  Textbasierte Suchanfragen können nicht nur über visuelle Ähnlichkeiten, sondern auch über die semantisch beschriebene Bedeutung der Bilder beantwortet werden.

* **Höhere Interpretierbarkeit:**
  Die generierte Bildbeschreibung ermöglicht eine transparente Nachvollziehbarkeit der zugrunde liegenden Repräsentationen und unterstützt bei der Evaluierung der Ergebnisse.

* **Gute Erweiterbarkeit:**
  Das Verfahren ist modular aufgebaut und lässt sich leicht um weitere Modalitäten wie Audio oder Video ergänzen.


<p><font color='black' size="5">
RAG-Prozess für Texte
</font></p>

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/rag_process.png" width="600" alt="Avatar">

<p><font color='black' size="5">
RAG-Prozess für Bilder
</font></p>

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/rag_process_03.png" width="600" alt="Avatar">

# 2 | Modul `multimodal_rag`
---

Python-Modul für ein **funktionales multimodales RAG-System**, das Text- und Bilddokumente in einer einheitlichen Vektordatenbank verwaltet und durchsucht.

**Modalitätsrichtungen**

| Eingabe (Query) | Ausgabe (Antwort) | Beispiel / Beschreibung | Status |
|-----------------|-------------------|-------------------------|--------|
| **Text → Text** | Textbasierte Frage führt zu Textantwort | Klassisches RAG-System (z.B. Chatbot, Q&A) | ✅ |
| **Text → Bild** | Textanfrage findet relevante Bilder | "Zeige mir Roboter-Bilder" | ✅ |
| **Bild → Text** | Bildanalyse oder Captioning | "Was ist auf diesem Foto zu sehen?" | ✅ |
| **Bild → Bild** | Bildretrieval oder visuelle Transformation | "Finde ähnliche Bilder" | ✅ |
| **Text + Bild → Text** | Kombination zur Textgenerierung | "Welche Informationen enthält dieses Diagramm?" | ❌ |
| **Text + Bild → Bild** | Bedingte Bildgenerierung | "Mach aus diesem Bild eine Winterversion" | ❌ |

**Hauptvorteile**

1. **Funktionale Architektur**: Klare Trennung von Konfiguration, Komponenten und Funktionen
2. **Einheitliche Datenbank**: ChromaDB mit separaten Collections für Text und Bilder
3. **Hybride Suche**: Text-Embeddings (OpenAI) + Bild-Embeddings (CLIP)
4. **Flexible Konfiguration**: Alle Parameter über RAGConfig anpassbar
5. **Bildbeschreibung**: Zu jedem Bild wird zusätzlich eine Bildbeschreibung erstellt.

<p><font color='black' size="5">
Aufbau
</font></p>

**RAGConfig**
- Zentrale Konfigurationsklasse
- Anpassbare Parameter (chunk_size, models, thresholds)

**RAGComponents**
- Container für alle System-Komponenten
- Text-Embeddings, CLIP-Model, LLMs, Collections

**Hauptfunktionen**
- Datensammlung
    - `init_rag_system()`: System-Initialisierung
    - `process_directory()`: Bulk-Import von Dateien
    - `add_text_document()`: Einzelnes Dokument hinzufügen
    - `add_image_with_description()`: Bild mit Auto-Beschreibung
- Abruf & Erweiterung
    - `search_texts()`: Text-Suche inkl. Bildbeschreibungen
    - `search_images()`: CLIP-basierte Bildsuche
    - `search_similar_images()`: Bild→Bild Ähnlichkeitssuche
    - `search_text_by_image()`: Bild→Text Suche
    - `multimodal_search()`: Erweiterte multimodale Suche

<img src="https://raw.githubusercontent.com/ralf-42/Image/main/multimodal_rag_function_call.png" width="1100" alt="Avatar">




In [4]:
# Import des Moduls
from multimodal_rag_modul import (
    init_rag_system,
    process_directory,
    multimodal_search,
    search_texts,
    search_images,
    search_text_by_image,
    search_similar_images,
    add_text_document,
    add_image_with_description,
    get_system_status
)

# 3 | RAG initialisieren
---

In [5]:
# Initialisierung
rag = init_rag_system()

🚀 Initialisiere Enhanced RAG-System in ./multimodal_rag_db
✅ OpenAI Text-Embeddings initialisiert
🖼️ Lade CLIP-Modell...


modules.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/604 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/389 [00:00<?, ?B/s]

0_CLIPModel/model.safetensors:   0%|          | 0.00/605M [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/316 [00:00<?, ?B/s]

0_CLIPModel/pytorch_model.bin:   0%|          | 0.00/605M [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.


✅ CLIP-Modell geladen
✅ LLMs initialisiert (Text + Vision)
✅ Collections initialisiert



In [6]:
# Konfiguration abfragen
print(rag.config)

# Konfiguration ändern (Beispiel: chunk_size ändern)
# rag.config.chunk_size = 300

RAGConfig(chunk_size=200, chunk_overlap=20, text_threshold=1.2, image_threshold=0.8, clip_model='clip-ViT-B-32', text_model='text-embedding-3-small', llm_model='gpt-4o-mini', vision_model='gpt-4o-mini', db_path='./multimodal_rag_db')


# 4 | Bilder/Texte in RAG übernehmen
---

In [7]:
# Dokumente/Bilder laden & verarbeiten
process_directory(rag, './files', auto_describe_images=True)

📊 Gefunden: 5 Text-Dateien, 3 Bilder

📄 biografien_3.pdf
✅ 68 Chunks von 'biografien_3.pdf' hinzugefügt
📄 roboter.txt
✅ 11 Chunks von 'roboter.txt' hinzugefügt
📄 biografien_2.md
✅ 44 Chunks von 'biografien_2.md' hinzugefügt
📄 biografien_1.txt
✅ 32 Chunks von 'biografien_1.txt' hinzugefügt
📄 biografien_4.docx
✅ 63 Chunks von 'biografien_4.docx' hinzugefügt

🖼️ hedra_cyborg.png
🖼️ Bild-Embedding erstellt für hedra_cyborg.png
📝 Bildbeschreibung generiert: Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metall...
✅ Bildbeschreibung in Text-Collection gespeichert
✅ Bild 'hedra_cyborg.png' mit Cross-References hinzugefügt

🖼️ apfel.png
🖼️ Bild-Embedding erstellt für apfel.png
📝 Bildbeschreibung generiert: Das Bild zeigt einen glänzenden roten Apfel, der auf einer Holzoberfläche platziert ist. Der Apfel h...
✅ Bildbeschreibung in Text-Collection gespeichert
✅ Bild 'apfel.png' mit Cross-References hinzugefügt

🖼️ retro_robot.jpg
🖼️ Bild-Embedding e

{'texts': 5, 'images': 3, 'image_descriptions': 3}

# 5 | Suche mit Text
---

<p><font color='black' size="5">
Text → Text
</font></p>

In [8]:
result = search_texts(rag, "Was weisst Du über Cyborgs?")
mprint(result)

[search_texts] 🤖 LLM-ANTWORT:
----------------------------------------------------------------------
Cyborgs sind hybride Wesen, die sowohl biologische als auch technologische Komponenten vereinen. Sie können Menschen sein, die durch mechanische oder elektronische Teile ergänzt wurden, um ihre Fähigkeiten zu erweitern oder zu verbessern. In der Science-Fiction werden Cyborgs oft als Charaktere dargestellt, die übermenschliche Fähigkeiten besitzen oder in der Lage sind, mit Maschinen zu interagieren. Ethische Fragen, die mit Cyborgs verbunden sind, betreffen unter anderem Identität, Verantwortung und die Auswirkungen der Technologie auf das menschliche Leben.

[search_texts] 📚 VERWENDETE QUELLEN:
----------------------------------------------------------------------

   📄 Text-Dokumente (2):
      1. roboter.txt (Ähnlichkeit: 0.444)
      2. roboter.txt (Ähnlichkeit: 0.431)

   🖼️ Bildbeschreibungen (1):
      1. hedra_cyborg.png (Ähnlichkeit: 0.529)


<p><font color='black' size="5">
Text → Bild
</font></p>

In [9]:
result = search_images(rag, "Was weisst Du über Cyborgs?")
for img in result:
    mprint(f"**🖼️ {img['filename']}**")
    mprint(img['description'])

**🖼️ hedra_cyborg.png**

Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische und strukturierte Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Die Komposition ist fokussiert auf das Gesicht und den Oberkörper des Charakters, während der Hintergrund unscharf bleibt und eine moderne Umgebung andeutet. Besondere Merkmale sind die detaillierte Textur des Anzugs und der entschlossene Ausdruck d

**🖼️ retro_robot.jpg**

Das Bild zeigt einen freundlichen, silbernen Roboter, der vor einem Gebäude steht, das mit einem Schild "WELCOME ROBOTS WELCOME" geschmückt ist. Die Farben sind hell und einladend, mit warmen Tönen, die eine positive und futuristische Stimmung erzeugen. Die Komposition zeigt den Roboter im Vordergrund, während im Hintergrund weitere Roboter sichtbar sind, was auf eine belebte Umgebung hinweist. Besondere Merkmale sind die glänzende Oberfläche des Roboters und die retro-futuristische Gestaltung d

<p><font color='black' size="5">
Text → Text/Bild
</font></p>

In [10]:
result = multimodal_search(rag, "Was weisst Du über Cyborgs?")
mprint(result)


[multimodal_search] ======================================================================
[multimodal_search] 🔍 Multimodale Suche: Was weisst Du über Cyborgs?
[multimodal_search] ======================================================================



🤖 LLM-ZUSAMMENFASSUNG:
======================================================================
Cyborgs sind hybride Wesen, die sowohl biologische als auch technologische Komponenten vereinen. Sie können Menschen sein, die durch mechanische oder elektronische Teile ergänzt oder verbessert wurden, um ihre Fähigkeiten zu erweitern oder bestimmte Funktionen zu erfüllen. In der Science-Fiction werden Cyborgs oft als Charaktere dargestellt, die übermenschliche Fähigkeiten besitzen oder in der Lage sind, mit Maschinen zu interagieren. Ethische Fragen, die mit Cyborgs verbunden sind, betreffen unter anderem Identität, Verantwortung und die Auswirkungen der Technologie auf das menschliche Leben.

📚 VERWENDETE QUELLEN (via semantische Textsuche):
======================================================================

   📄 Text-Dokumente (2):
      1. roboter.txt (Ähnlichkeit: 0.444)
      2. roboter.txt (Ähnlichkeit: 0.43)

   🖼️ Bildbeschreibungen (1):
      1. hedra_cyborg.png (Ähnlichkeit: 0.529)

🔍 BILDER VIA CLIP (direkte visuelle Ähnlichkeitssuche):
======================================================================
   1. hedra_cyborg.png (Ähnlichkeit: 0.239)
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische und strukturierte Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Die Komposition ist fokussiert auf das Gesicht un...
   2. retro_robot.jpg (Ähnlichkeit: 0.22)
      📝 Das Bild zeigt einen freundlichen, silbernen Roboter, der vor einem Gebäude steht, das mit einem Schild "WELCOME ROBOTS WELCOME" geschmückt ist. Die Farben sind hell und einladend, mit warmen Tönen, die eine positive und futuristische Stimmung erzeugen. Die Komposition zeigt den Roboter im Vordergru...

🔗 BILDER VIA CROSS-MODAL RETRIEVAL (Text → Bild Verknüpfung):
======================================================================
   1. hedra_cyborg.png
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische und strukturierte Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Die Komposition ist fokussiert auf das Gesicht un...


# 6 | Suche mit Bild
---

<p><font color='black' size="5">
Bild → Bild
</font></p>

In [11]:
# Bild → Bild: Finde visuell ähnliche Bilder
result = search_similar_images(rag, "./zwei_roboter.png", k=5)
mprint("## 🖼️ Suche Bild → Bild")
mprint("---")
for img in result:
    mprint(f"{img['filename']}: Ähnlichkeit: {img['similarity']}")

[search_similar_images] 🔍 CLIP-Embedding erstellt für Query-Bild: zwei_roboter.png
[search_similar_images] ✅ 3 ähnliche Bilder gefunden



## 🖼️ Suche Bild → Bild

---

retro_robot.jpg: Ähnlichkeit: 0.829

apfel.png: Ähnlichkeit: 0.677

hedra_cyborg.png: Ähnlichkeit: 0.469

<p><font color='black' size="5">
Bild → Text
</font></p>

In [12]:
# Bild → Text: Finde Textinformationen zu dem Bild
result = search_text_by_image(rag, "./zwei_roboter.png", k=5)
mprint(result)


[search_text_by_image] ======================================================================
[search_text_by_image] 🔍 Bild → Text Suche für: zwei_roboter.png
[search_text_by_image] ======================================================================

[search_similar_images] 🔍 CLIP-Embedding erstellt für Query-Bild: zwei_roboter.png
[search_similar_images] ✅ 3 ähnliche Bilder gefunden

[search_text_by_image] 🔍 Suche nach relevanten Text-Dokumenten basierend auf Bildbeschreibungen...
[search_text_by_image] ✅ 3 relevante Text-Dokumente gefunden



[search_text_by_image] 🤖 LLM-ZUSAMMENFASSUNG:
======================================================================
Die Analyse des Query-Bildes zeigt eine starke visuelle und semantische Verbindung zu Robotern und futuristischen Themen. 

**Visuelle Ähnlichkeiten:**
1. **Retro-Roboter**: Das Bild eines freundlichen, silbernen Roboters vor einem einladenden Gebäude mit dem Schild "WELCOME ROBOTS WELCOME" hebt die positive und futuristische Stimmung hervor. Die glänzende Oberfläche und die retro-futuristische Gestaltung deuten auf eine Umgebung hin, in der Roboter akzeptiert und integriert sind.
2. **Cyborg-Charakter**: Der männliche Charakter in einem futuristischen Anzug vermittelt eine ernste und kraftvolle Atmosphäre, die oft in Science-Fiction-Settings zu finden ist. Die detaillierte Textur und der entschlossene Ausdruck des Charakters betonen die Verbindung zu technologischen und futuristischen Themen.

**Relevante Textdokumente:**
Die Dokumente über Roboter verdeutlichen, dass Roboter nicht nur technische Werkzeuge sind, sondern auch Ausdruck menschlicher Kreativität und Ambitionen. Sie sind mittlerweile ein fester Bestandteil des Alltags und der industriellen Produktion, wobei sie Aufgaben übernehmen, die für Menschen gefährlich oder monoton sind. Die Entwicklung von Servicerobotern und kollaborativen Robotern zeigt, dass sich das Bild des Roboters weiterentwickelt und sie zunehmend in verschiedenen Lebensbereichen integriert werden.

**Gemeinsamkeiten und wichtige Merkmale:**
- **Technologische Integration**: Sowohl die Bilder als auch die Texte thematisieren die Integration von Robotern in die Gesellschaft und deren Rolle in der Zukunft.
- **Futuristische Ästhetik**: Die visuelle Darstellung der Roboter und Cyborgs spiegelt eine futuristische Ästhetik wider, die in den Texten als Teil der menschlichen Ambitionen beschrieben wird.
- **Positive Wahrnehmung**: Die freundliche Darstellung des Roboters und die einladende Umgebung deuten auf eine positive Wahrnehmung von Robotern hin, die in den Texten als zunehmend akzeptierte und nützliche Begleiter beschrieben werden.

Insgesamt zeigen die Informationen, dass Roboter und futuristische Technologien nicht nur technische Innovationen sind, sondern auch kulturelle Symbole für Fortschritt und menschliche Kreativität darstellen.

[search_text_by_image] 🔍 VERWENDETE QUELLEN (visuell ähnliche Bilder via CLIP):
======================================================================
   1. retro_robot.jpg (Ähnlichkeit: 0.829)
      📝 Das Bild zeigt einen freundlichen, silbernen Roboter, der vor einem Gebäude steht, das mit einem Schild "WELCOME ROBOTS WELCOME" geschmückt ist. Die Farben sind hell und einladend, mit warmen Tönen, die eine positive und futuristische Stimmung erzeugen. Die Komposition zeigt den Roboter im Vordergru...
   2. apfel.png (Ähnlichkeit: 0.677)
      📝 Das Bild zeigt einen glänzenden roten Apfel, der auf einer Holzoberfläche platziert ist. Der Apfel hat eine frische, lebendige Farbe und wird durch ein grünes Blatt ergänzt, das ihm eine natürliche Note verleiht. Die Komposition ist einfach und fokussiert, mit einem neutralen, hellen Hintergrund, de...
   3. hedra_cyborg.png (Ähnlichkeit: 0.469)
      📝 Das Bild zeigt einen männlichen Charakter in einem futuristischen, eng anliegenden Anzug, der metallische und strukturierte Elemente aufweist. Die dominierenden Farben sind dunkles Blau und Schwarz, was eine ernste und kraftvolle Stimmung vermittelt. Die Komposition ist fokussiert auf das Gesicht un...

[search_text_by_image] 📚 RELEVANTE TEXT-DOKUMENTE (via semantische Suche):
======================================================================
   1. roboter.txt (Ähnlichkeit: 0.521)
      📄 Roboter sind damit nicht nur technische Werkzeuge, sondern Ausdruck menschlicher Kreativität und Ambition – sie spiegeln wider, wie wir uns selbst, unsere Arbeit und unsere Zukunft verstehen. Der...
   2. roboter.txt (Ähnlichkeit: 0.482)
      📄 kann. Gleichzeitig verändert sich das Bild des Roboters zunehmend: Neben den klassischen Industrierobotern treten heute Serviceroboter, autonome Fahrzeuge und kollaborative Roboter („Cobots“) auf,...
   3. roboter.txt (Ähnlichkeit: 0.426)
      📄 Roboter sind längst keine Zukunftsvision mehr, sondern fester Bestandteil unseres Alltags und der industriellen Produktion. Sie übernehmen Aufgaben, die für Menschen zu gefährlich, monoton oder...


# 7 | Aufgabe
---

<p><font color='black' size="5">
Teste das Multimodale RAG-System mit eigenen Queries
</font></p>

**Ziel**: Nutze das fertige multimodale RAG-System, um verschiedene Such-Operationen durchzuführen.



**Schritt 1: Einfache Queries testen**

Führe die folgenden Queries mit dem RAG-System aus und beobachte, welche Ergebnisse zurückgeliefert werden:

```python
# Query 1: Text → Text (Klassisches RAG)
result = multimodal_search(rag, "Wer war Alan Turing?")
mprint(result)

# Query 2: Text → Bild (Textanfrage findet Bilder)
result = multimodal_search(rag, "Zeige mir Bilder von Robotern")
mprint(result)

# Query 3: Bild → Bild (Finde ähnliche Bilder)
similar_images = search_similar_images(rag, "./files/apfel.jpg", k=3)
mprint("## 🖼️ Ähnliche Bilder:")
for img in similar_images:
    mprint(f"  • {img['filename']}: Ähnlichkeit {img['similarity']:.2f}")

# Query 4: Bild → Text (Finde Textinformationen zu einem Bild)
text_result = search_text_by_image(rag, "./files/hedra_cyborg.png", k=3)
mprint(text_result)
```



**Schritt 2: Eigene Queries erstellen**

Entwickle mindestens **3 eigene Queries**, die verschiedene Modalitäten nutzen:

**Beispiele für kreative Queries:**

- **Text→Text**: "Was ist der Unterschied zwischen Robotern und Cyborgs?"
- **Text→Bild**: "Finde alle futuristischen Bilder"
- **Bild→Bild**: Suche ähnliche Bilder zu `a_retro-futuristic_robot_dall_e.jpg`
- **Bild→Text**: Finde Textinformationen, die zum Apfel-Bild passen



**Schritt 3: Bonus-Challenge (Optional)**

Experimentiere mit den Suchparametern:

```python
# Ändere die Anzahl der Ergebnisse (k)
result = multimodal_search(rag, "Roboter", k=5)

# Vergleiche verschiedene Bilder für die Bild→Bild Suche
similar_1 = search_similar_images(rag, "./files/apfel.jpg", k=3)
similar_2 = search_similar_images(rag, "./files/hedra_cyborg.png", k=3)

# Finde heraus, welche Bilder am ähnlichsten zueinander sind
```

# A | Ähnlichkeitsmessung
---


<p><font color='black' size="5">
1. Text-Ähnlichkeit (semantisch)
</font></p>

  Embedding-Modell: OpenAI text-embedding-3-small

  Messmethode:
  - ChromaDB nutzt L2-Distanz (Euklidische Distanz) für Text-Embeddings
  - Werte: 0 = identisch, 2 = maximal entfernt
  - Konvertierung Distanz zu Ähnlichkeit: similarity = max(0, 1 - (score / 2))
  - Threshold: text_threshold: 1.2 (Zeile 51), Mindest-Ähnlichkeit: 0.3

  Verwendung in:
  - search_texts() - Text-Dokumente und Bildbeschreibungen durchsuchen
  - multimodal_search() - Kombinierte Suche

  ---

<p><font color='black' size="5">
2. Bild-Ähnlichkeit (visuell)
</font></p>

  Embedding-Modell: CLIP clip-ViT-B-32

  Messmethode:
  - ChromaDB nutzt Cosine-Distanz für Bild-Embeddings (Zeile 131: "hnsw:space": "cosine")
  - Werte: 0 = identisch, 2 = maximal entfernt
  - Konvertierung zu Ähnlichkeit:
  similarity = round(max(0, 1 - distance), 3)
  - Threshold: image_threshold: 0.8

  Verwendung in:
  - search_images() - Text → Bild Suche über CLIP   
  - search_similar_images() - Bild → Bild Suche  

  ---
  

<p><font color='black' size="5">
3. Cross-Modal-Retrieval (Text ↔ Bild)
</font></p>


  Methode: Indirekte Ähnlichkeit über Bildbeschreibungen

  Ablauf:
  1. Text → Bild: Text-Suche findet Bildbeschreibungen → verknüpfte Bilder werden abgerufen  
  2. Bild → Text: Bild-Suche findet ähnliche Bilder → deren Beschreibungen werden für semantische Textsuche verwendet

  Verknüpfung: Beide Collections sind über text_doc_id ↔ image_doc_id referenziert

  ---
  

<p><font color='black' size="5">
Zusammenfassung der Metriken:
</font></p>

  | Modalität | Embedding-Modell              | Distanzmetrik  | Threshold | Ähnlichkeitsbereich |
  |-----------|-------------------------------|----------------|-----------|---------------------|
  | Text      | OpenAI text-embedding-3-small | L2-Distanz     | 1.2       | 0.3 - 1.0           |
  | Bild      | CLIP ViT-B-32                 | Cosine-Distanz | 0.8       | 0.0 - 1.0           |

  Die Konvertierung 1 - (distance / 2) normalisiert beide Distanzmaße auf einen Ähnlichkeitswert von 0 bis 1, wobei 1 = maximale Ähnlichkeit bedeutet.